test/tstDetailsTreeNodeLabeling.cpp

```diff
@@ -0,0 +1,206 @@
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsTreeNodeLabeling.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+#include <random>
+
+template <typename DeviceType, typename T>
+auto toView(std::vector<T> const &v, std::string const &lbl = "")
+{
+  Kokkos::View<T *, DeviceType> view(lbl, v.size());
+  Kokkos::deep_copy(view, Kokkos::View<T const *, Kokkos::HostSpace,
+                                       Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                              v.data(), v.size()));
+  return view;
+}
+
+template <class MemorySpace>
+struct MockBVH
+{
+  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> children_;
+  KOKKOS_FUNCTION auto size() const { return children_.extent(0) + 1; }
+};
+
+#define HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(MEMORY_SPACE)           \
+  template <>                                                                  \
+  KOKKOS_FUNCTION auto                                                         \
+  ArborX::Details::HappyTreeFriends::getLeftChild<MockBVH<MEMORY_SPACE>>(      \
+      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
+  {                                                                            \
+    return x.children_(i).first;                                               \
+  }                                                                            \
+  template <>                                                                  \
+  KOKKOS_FUNCTION auto                                                         \
+  ArborX::Details::HappyTreeFriends::getRightChild<MockBVH<MEMORY_SPACE>>(     \
+      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
+  {                                                                            \
+    return x.children_(i).second;                                              \
+  }
+
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::HostSpace)
+#ifdef KOKKOS_ENABLE_CUDA
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::Cuda::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_HIP
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::HIP::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_SYCL
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::SYCL::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::OpenMPTarget::memory_space)
+#endif
```

Comment:
We would need to conditionally specialize on `HostSpace` if they are not the same but yes this could read
```C++
HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::DefaultExecutionSpace::memory_space)
#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) || defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ENABLE_OPENMPTARGET)
static_assert(!std::is_same_v<Kokkos::DefaultExecutionSpace::memory_space, Kokkos::DefaultHostExecutionSpace::memory_space>);
HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::DefaultHostExecutionSpace::memory_space)
#endif

Reply:
Do you want me to change?

Reply:
I think it's fine.